crates/blockifier/src/execution/contract_class.rs

```diff
@@ -681,18 +666,29 @@ impl NativeContractClassV1Inner {
                 &lookup_fid,
                 &sierra_contract_class.entry_points_by_type,
             )?,
-            sierra_program_raw: sierra_contract_class.sierra_program,
-            fallback_entry_points_by_type: sierra_contract_class.entry_points_by_type,
+            // sierra_program_hash : calculate_block_hash(sierra_contract_class.sierra_program)
+// todo: add hash of progam here for PoC
+            sierra_program_hash : calculate_sierra_program_hash(sierra_contract_class.sierra_program),
+            // sierra_program_raw: sierra_contract_class.sierra_program,
+            // fallback_entry_points_by_type: sierra_contract_class.entry_points_by_type,
         })
     }
 }
 
+fn calculate_sierra_program_hash(sierra: Vec<BigUintAsHex>) -> BlockHash {
+  let sierra_felts : Vec<Felt> = sierra.iter().map(|wrapper| &wrapper.value).map_into().collect();
+  BlockHash(
+    HashChain::new()
+    .chain_size_and_elements(&sierra_felts)
+        .get_poseidon_hash())
+}
+
 // The location where the compiled contract is loaded into memory will not
 // be the same therefore we exclude it from the comparison.
 impl PartialEq for NativeContractClassV1Inner {
     fn eq(&self, other: &Self) -> bool {
         self.entry_points_by_type == other.entry_points_by_type
```

Comment:
Do we still need this comparison?

Reply:
Yes. I've added an example as to why under the `crates/blockifier/tests/test_native_contract.rs`.
`test_partial_eq` will pass on the current `native2.7.x`
`test_partial_eq` will pass on the current version of this branch (`pwhite/remove_fallback`)
`test_partial_eq` will fail on this branch if `self.entry_points_by_type == other.entry_points_by_type` is removed.

Under `crates/blockifier/feature_contracts/cairo1/compiled/`, if we look at `test_contract_entrypoint_a.sierra.json` and `test_contract_entrypoint_b.sierra.json`, we can see that the `sierra_program` program is the same for both files, but the `entry_points_by_type` is different.

Under `crates/blockifier/feature_contracts/cairo1/`, the `.cairo` files `test_contract_entrypoint_a.cairo` and `test_contract_entrypoint_b.cairo` are included to show that we can indeed create contracts with identical sierra code but different entry points. Note that this is only the case for different function names with the same code as even changing the return value (arbitrarily set to 73) would result in a change in the sierra program no matter how you order the functions.